nemo/collections/tts/modules/spectrogram_enhancer.py

```diff
@@ -0,0 +1,359 @@
+        res = self.conv_res(x)
+        x = self.net(x)
+        if self.downsample is not None:
+            x = self.downsample(x)
+        x = (x + res) * (1 / math.sqrt(2))
+        return x
+
+
+class Generator(torch.nn.Module):
+    def __init__(
+        self, n_bands, latent_dim, style_depth, network_capacity=16, channels=1, fmap_max=512,
+    ):
+        super().__init__()
+        self.image_size = n_bands
+        self.latent_dim = latent_dim
+        self.num_layers = int(log2(n_bands) - 1)
+        self.style_depth = style_depth
+
+        self.style_mapping = StyleMapping(self.latent_dim, self.style_depth, lr_mul=0.1)
+
+        filters = [network_capacity * (2 ** (i + 1)) for i in range(self.num_layers)][::-1]
+
+        set_fmap_max = partial(min, fmap_max)
+        filters = list(map(set_fmap_max, filters))
+        init_channels = filters[0]
+        filters = [init_channels, *filters]
+
+        in_out_pairs = zip(filters[:-1], filters[1:])
+        initial_block_size = n_bands // 2 ** (self.num_layers - 1), 800
+        self.initial_block = torch.nn.Parameter(
+            torch.randn((1, init_channels, *initial_block_size)), requires_grad=False
+        )
+
+        self.initial_conv = torch.nn.Conv2d(filters[0], filters[0], 3, padding=1)
+        self.blocks = torch.nn.ModuleList([])
+
+        for ind, (in_chan, out_chan) in enumerate(in_out_pairs):
+            not_first = ind != 0
+            not_last = ind != (self.num_layers - 1)
+
+            block = GeneratorBlock(
+                latent_dim, in_chan, out_chan, upsample=not_first, upsample_rgb=not_last, channels=channels,
+            )
+            self.blocks.append(block)
+
+        for m in self.modules():
+            if isinstance(m, (torch.nn.Conv2d, torch.nn.Linear)):
+                torch.nn.init.kaiming_normal_(m.weight, a=0, mode="fan_in", nonlinearity="leaky_relu")
+        for block in self.blocks:
+            torch.nn.init.zeros_(block.to_noise1.weight)
+            torch.nn.init.zeros_(block.to_noise1.bias)
+            torch.nn.init.zeros_(block.to_noise2.weight)
+            torch.nn.init.zeros_(block.to_noise2.bias)
+
+    def add_scaled_condition(self, target: torch.Tensor, condition: torch.Tensor, condition_lengths: torch.Tensor):
+        *_, target_height, _ = target.shape
+        *_, height, _ = condition.shape
+
+        scale = height // target_height
+
+        # scale appropriately
+        condition = F.interpolate(condition, size=target.shape[-2:], mode="bilinear")
+
+        # add and mask
+        result = (target + condition) / 2
+        result = mask_sequence_tensor(result, (condition_lengths / scale).ceil().long())
+
+        return result
+
+    def forward(self, condition: torch.Tensor, lengths: torch.Tensor, ws: List[torch.Tensor], noise: torch.Tensor):
+        batch_size, _, _, max_length = condition.shape
+
+        x = self.initial_block.expand(batch_size, -1, -1, -1)
+        x = x[:, :, :, : max_length // 16]
```

Comment:
what is 16?

Reply:
This depends on the fact that we have four 2x upsample steps, so 2**4

Supposed to depend on number of blocks, also a good catch

Reply:
Done, made a computable property for this